pkg/runtime/types.go

```diff
@@ -120,26 +132,370 @@ type Unknown struct {
+				strMap[k] = str
+			}
+		}
+		return strMap
+	}
+	return nil
+}
+
+func setNestedField(obj map[string]interface{}, value interface{}, fields ...string) {
+	m := obj
+	if len(fields) > 1 {
+		for _, field := range fields[0 : len(fields)-1] {
+			if _, ok := m[field].(map[string]interface{}); !ok {
+				m[field] = make(map[string]interface{})
+			}
+			m = m[field].(map[string]interface{})
+		}
+	}
+	m[fields[len(fields)-1]] = value
+}
+
+func setNestedMap(obj map[string]interface{}, value map[string]string, fields ...string) {
+	m := make(map[string]interface{}, len(value))
+	for k, v := range value {
+		m[k] = v
+	}
+	setNestedField(obj, m, fields...)
+}
+
+func (u *Unstructured) setNestedField(value interface{}, fields ...string) {
+	if u.Object == nil {
+		u.Object = make(map[string]interface{})
+	}
+	setNestedField(u.Object, value, fields...)
+}
+
+func (u *Unstructured) setNestedMap(value map[string]string, fields ...string) {
+	if u.Object == nil {
+		u.Object = make(map[string]interface{})
+	}
+	setNestedMap(u.Object, value, fields...)
+}
+
+// TODO: rewrite this.
+func setField(src reflect.Value, fieldName string, dest interface{}) error {
+	field := src.FieldByName(fieldName)
+	if !field.IsValid() {
+		return fmt.Errorf("%v does not have field %s", src, fieldName)
+	}
+	v, err := conversion.EnforcePtr(dest)
+	if err != nil {
+		return err
+	}
+	if field.Type().AssignableTo(v.Type()) {
+		v.Set(field)
+		return nil
+	}
+	if field.Type().ConvertibleTo(v.Type()) {
+		v.Set(field.Convert(v.Type()))
+		return nil
+	}
+	return fmt.Errorf("couldn't assign/convert %v to %v", field.Type(), v.Type())
+}
+
+func extractOwnerReference(src interface{}) (metatypes.OwnerReference, error) {
+	var ret metatypes.OwnerReference
+	v := src.(map[string]interface{})
+	ret.Kind = getNestedString(v, "kind")
+	ret.Name = getNestedString(v, "name")
+	ret.APIVersion = getNestedString(v, "apiVersion")
+	ret.UID = (types.UID)(getNestedString(v, "uid"))
+	return ret, nil
+}
+
+func setOwnerReference(dest interface{}, src metatypes.OwnerReference) error {
+	v, ok := dest.(map[string]interface{})
+	if !ok {
+		return fmt.Errorf("dest %v has to be of type map[string]interface()", dest)
+	}
+	setNestedField(v, src.Kind, "kind")
+	setNestedField(v, src.Name, "name")
+	setNestedField(v, src.APIVersion, "apiVersion")
+	setNestedField(v, string(src.UID), "uid")
+	return nil
+}
+
+func getOwnerReferences(object map[string]interface{}) (ownerReferences []map[string]interface{}, err error) {
+	field := getNestedField(object, "metadata", "ownerReferences")
+	if field == nil {
+		return ownerReferences, fmt.Errorf("cannot find field metadata.ownerReferences in %v", object)
+	}
+	ownerReferences, ok := field.([]map[string]interface{})
+	if !ok {
+		return ownerReferences, fmt.Errorf("expect metadata.ownerReferences to be a slice in %v", object)
+	}
+	return
+}
+
+func (u *Unstructured) GetOwnerReference(i int) (metatypes.OwnerReference, error) {
```

Comment:
I added benchmark for Append and Remove operations, as I think they will have larger difference.
https://play.golang.org/p/yFYzuIs69M

Result:
tl;dr, when there are 10 references, with copy the Remove is 9x slower, Get is almost the same, Append is 50x slower.
```
          AccessAllDirect(  1): 10000000	       141 ns/op
            AccessAllCopy(  1):  5000000	       239 ns/op
             RemoveDirect(  1): 50000000	        25.2 ns/op
           RemoveWithCopy(  1):  5000000	       354 ns/op
             AppendDirect(  1): 20000000	        92.7 ns/op
           AppendWithCopy(  1):  1000000	      1839 ns/op
--------------
          AccessAllDirect(  2):  5000000	       276 ns/op
            AccessAllCopy(  2):  3000000	       429 ns/op
             RemoveDirect(  2): 10000000	       151 ns/op
           RemoveWithCopy(  2):  1000000	      1275 ns/op
             AppendDirect(  2): 10000000	       137 ns/op
           AppendWithCopy(  2):   500000	      2818 ns/op
--------------
          AccessAllDirect(  5):  2000000	       651 ns/op
            AccessAllCopy(  5):  2000000	       964 ns/op
             RemoveDirect(  5):  5000000	       295 ns/op
           RemoveWithCopy(  5):   500000	      3870 ns/op
             AppendDirect(  5): 10000000	       166 ns/op
           AppendWithCopy(  5):   300000	      5541 ns/op
--------------
          AccessAllDirect( 10):  1000000	      2053 ns/op
            AccessAllCopy( 10):  1000000	      2010 ns/op
             RemoveDirect( 10):  1000000	      1013 ns/op
           RemoveWithCopy( 10):   200000	      9133 ns/op
             AppendDirect( 10): 10000000	       194 ns/op
           AppendWithCopy( 10):   200000	     10302 ns/op
--------------
          AccessAllDirect( 20):   500000	      2552 ns/op
            AccessAllCopy( 20):   500000	      3574 ns/op
             RemoveDirect( 20):  1000000	      1301 ns/op
           RemoveWithCopy( 20):   100000	     17177 ns/op
             AppendDirect( 20):  5000000	       242 ns/op
           AppendWithCopy( 20):   100000	     18950 ns/op
--------------
          AccessAllDirect( 50):   200000	      6316 ns/op
            AccessAllCopy( 50):   200000	      8709 ns/op
             RemoveDirect( 50):   500000	      3202 ns/op
           RemoveWithCopy( 50):    30000	     45076 ns/op
             AppendDirect( 50):  3000000	       436 ns/op
           AppendWithCopy( 50):    30000	     44644 ns/op
--------------
          AccessAllDirect(100):   100000	     13982 ns/op
            AccessAllCopy(100):   100000	     17911 ns/op
             RemoveDirect(100):   200000	      6389 ns/op
           RemoveWithCopy(100):    20000	     95147 ns/op
             AppendDirect(100):  2000000	       687 ns/op
           AppendWithCopy(100):    20000	     97957 ns/op
--------------
```